### PR TITLE
feat: l1 to l2 erc20 deposit

### PR DESCRIPTION
**Pull request by vasyl-ivanchuk**
L1 to L2 ERC20 deposit functionality. Tested using dockerized node locally and on Sepolia.

Example of the L1 deposit transaction:
https://sepolia.etherscan.io/tx/0xa48b24f46edd799982b1eb73201cf94ed625e7732f966d515a3faf10b6ee0cea

Corresponding L2 transaction:
https://sepolia.explorer.zksync.io/tx/0x55516f8c3c3290455637ce058acbd158b37d45022192e149bd79e759f893fdd4

### CODE REVIEW

**Comment by vasyl-ivanchuk** [comment]
@popzxc I will add a CI action to test deposits with the dockerized node, but in a separate PR.

File: src/provider/mod.rs
Comment (popzxc) [inline src/provider/mod.rs:597]: Given that we're about to get a few more options here (deposit custom token, deposit ERC20 to a custom base token chain), probably it makes sense to extract this logic into a separate function. The function probably shouldn't live on the `ZksyncProviderWithWallet` trait (to not create too many public methods here).

The way [it's currently implemented in zksync-ethers](https://github.com/zksync-sdk/zksync-ethers/blob/main/src/adapters.ts#L416) makes sense historically (as the number of supported combinations has grown) but doesn't feel maintainable, so I think we need different abstractions here.

We have several logical steps here:
1. Approve required tokens if needed. We may have approve either 0, 1, or 2 tokens:
  - Deposit ETH to ETH chain -> 0 tokens
  - Deposit ERC20 to ETH chain -> 1 token
  - Deposit native ERC20 to ERC20 chain -> 1 token
  - Deposit non-native ERC20 to ERC20 chain -> 2 tokens
2. Prepare tx data for L1. There are two options here:
  - `requestL2TransactionDirect` for native tokens
  - `requestL2TransactionTwoBridges` for non-native tokens (it also includes depositing ETH to non-ETH based token)
3. Set remaining fields (like `fee_per_gas`) and submit transaction. This is shared for all the tokens.
4. Get priority op response. This is shared for everyone.

What I'm thinking currently is something like this (pseudocode):

```rust
pub struct DepositExecutor<'a> {
  l1_provider: &'a L1Provider,
  l2_provider: &'a L2Provider,
  request: &'a DepositRequest
}

impl<'a> DepositExecutor<'a> {
  pub async fn execute(self) -> ... {
    let l1_fee_params = self.get_l1_fee_params().await?;
    // Get the final transaction to be submitted. We want to do it before
    // approving to decrease chances of deposit failing _after_ submitting any txs.
    let l1_deposit_tx = self.encode_deposit_calldata(&l1_fee_params).await?;
    // Approve required amount of tokens if needed. Can submit from 0 to 2 transactions.
    self.approve_tokens(&l1_fee_params).await?;
    // Now everything is ready, submit the deposit.
    self.submit(l1_deposit_tx).await
  }
}
```

Then the deposit function will be as simple as:

```rust
async fn deposit(&self, deposit_request: &DepositRequest, l1_provider: &P) -> ... {
  let executor = DepositExecutor { 
    l1_provider,
    l2_provider: self,
    request: deposit_request
  };
  executor.execute().await
}
```

Additional benefit here is that we will be able to later extract parts of it (e.g. provide an interface to just encode deposit calldata or just approve enough tokens)
easily without creating too much copy-paste.

File: src/provider/mod.rs
Comment (popzxc) [inline src/provider/mod.rs:630]: I think `DepositRequest` should have a setting to either disable automatic approvals, smth like 

```
impl DepositRequest {
  /// Enable or disable automatic submission of ERC20 approval transactions
  /// if the allowance is not sufficient.
  fn with_auto_approval(auto_approval: bool) -> { .. }
}
```

Later we will probably want to make it possible to set custom amounts for approval, but it's not needed right now.

**Review by popzxc — COMMENTED** [review]
Good progress here! Left a couple of design comments.

File: src/provider/mod.rs
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: Agreed, added the `auto_approval` option and the corresponding method to set it.

File: src/provider/mod.rs
Comment (vasyl-ivanchuk) [reply src/provider/mod.rs]: I added the `DepositExecutor` abstraction and the provider looks much cleaner. The `execute` method in the `DepositExecutor` is not as lean as in you pseudocode as we need to share some common state among the methods inside (l2 chain id, sender, receiver etc). Please, take a look.

**Review by popzxc — APPROVED** [review]
Nice! The fact that `DepositExecutor` is not as lean is expected, and since it already provides enough encapsulation, we can deal with further refactoring later :)

**Pull request merged**
